### PR TITLE
KAFKA-10556: NPE if sasl.mechanism is unrecognized

### PR DESCRIPTION
**Pull request by rondagostino**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java:217]: Should we fix server creation as well?

**Review by rajinisivaram — COMMENTED** [review]
@rondagostino Thanks for the PR, looks good. Left one comment.

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java
Comment (rondagostino) [reply clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java:217]: @rajinisivaram good point.  It's now fixed.  All client tests pass locally.

**Review by rajinisivaram — APPROVED** [review]
@rondagostino Thanks for the updates, LGTM. Will merge when builds complete.

**Comment by rajinisivaram** [comment]
Test failures look like flaky tests, there are not related to the PR. Merging to trunk.

**Comment by chia7712** [comment]
@rondagostino It seems ```MetricsTest.testMetrics```  start to fail after this PR is merged. Could you take a look?

**Comment by rondagostino** [comment]
Yeah, we no longer contact the server if the client doesn't recognize the SASL mechanism.  The PR at https://github.com/apache/kafka/pull/9372 fixes this test, which was using an unknown SASL mechanism (SCRAM-SHA-256) according to the JAAS config (PLAIN).  Now it will use PLAIN and bad credentials.

**Pull request merged**
